### ISSUE TITLE
Create a static method to pretty print gherkin

### ISSUE THREAD

**Issue by nddipiazza** [cucumber/common#1662]
I find myself wishing there was a simple SDK with the ability to pretty print (format) gherkin files.

Like how intellij cucumber plugin formats them is exactly what I want. 

Is there some way to add a simple static method for this? 

`GherkinFormatter.prettyPrint(gherkinString)`

I think there is a way to do this with the library, but it's not as simple.

**Comment by mpkorstanje** [comment]
I've moved this to the common repo where gherkin is maintained.

**Comment by aurelien-reeves** [comment]
There is a `pretty` method implemented in javascript. Would that fit your needs?

There is also https://github.com/cucumber/common/pull/1555 which work in enhancing that `pretty` method.

**Comment by nddipiazza** [comment]
I had originally opened this ticket in the cucumber-jvm project. it's there that I really need this. this was moved to the generic cucumber project which loses that fact.

**Comment by aurelien-reeves** [comment]
I considered the fact that you opened it in cucumber-jvm first.
Maybe the javascript version for gherkin-utils would had fit your needs

**Comment by aurelien-reeves** [comment]
Did you consider the possibility to use cucumber-jvm to get the messages in the form of ndjson to use those as an input in javascript implementation of gherkin and then get the gherkinDocument to use with gherkin-utils?

**Comment by mpkorstanje** [comment]
@nddipiazza what are you trying to achieve. You've described a solution, but not your problem.

**Comment by nddipiazza** [comment]
@mpkorstanje I have a java program that processes gherkin files. 
A portion of my java program needs to be able to, given a gherkin string, source format that gherkin. Most importantly so that the Test Data table is formatted pretty. 
Can this be done simply?

**Comment by mpkorstanje** [comment]
Is there a reason it has to be done in Java? 

We currently don't have any such function in Java. We do have a function that does it in Javascript. If it absolutely has to be Java then I guess you could port it. The structure isn't too different.

https://github.com/cucumber/common/blob/main/gherkin-utils/javascript/src/pretty.ts

**Comment by nddipiazza** [comment]
Yeah unfortunately in my case java is a requirement because my app server is java and there is no js engine available.

**Comment by aurelien-reeves** [comment]
I think that could be nice to port gherkin-utils to other languages. That's why I've added the "Help wanted" badge :)

**Comment by nddipiazza** [comment]
I could take a stab at it. there's not much code there. But i'm super weak with ts. So it'll take me a while

**Comment by laeubi** [comment]
Would be helpfull for IDEs as well!

**Comment by aurelien-reeves** [comment]
Short reminder: if anyone would need any kind of help, feel free to join our [community slack](https://cucumberbdd-slack-invite.herokuapp.com/). You'll find plenty of developers ready to help.

If for any reason you would not or can not join slack, we will find other way to provide any kind of help. It can be as part of this issue, or a pull-request, or something else? Are you using some tools which you would keep using? We can certainly join you there. Really, feel free to ask! We're here to help :smiley:

**Comment by aslakhellesoy** [comment]
We now have a `pretty(GherkinDocument) -> string` function [implemented in TypeScript](https://github.com/cucumber/common/blob/main/gherkin-utils/javascript/src/pretty.ts).

For implementations in other languages, please make a `gherkin-utils/[language]` implementation based on the same design (using a `walkGherkinDocument` function).

**Comment by nddipiazza** [comment]
@aslakhellesoy here you go - https://github.com/cucumber/common/pull/1725

**Comment by aslakhellesoy** [comment]
Great stuff @nddipiazza - let's continue the discussion in #1725

**Comment by jamietanna** [comment]
FYI we're making good progress with this, but there are still a few areas I think we should improve on before we can say this is done - see #2024 for some Gherkin we'd want to finalise the prety-printing for before saying this is "ready"
